### ISSUE TITLE
Blocked By

### ISSUE THREAD

**Issue by petey** [screwdriver-cd/screwdriver#653]
## Context

Build execution start should be able to be blocked by execution of another build, such that one build should not start until the other is completed.

## Objective

- [x] Jobs are always blocked by themselves. A build for any given job should not run concurrently with themself.
- [x] Jobs can be configured to be blocked by other jobs in the same pipeline

- [x] Jobs can be configured to be blocked by jobs in other pipelines.

- [x] UI should show a blocked task in status Blocked with reason. Reason should link to the running job which is blocking current job.

**Comment by d2lam** [comment]
# Proposed implementations
Using an example: B is blocked by A

## First way:
- Pass in the `blocked by` field to `executor start`
- Create a new plugin like [JobLock](https://github.com/taskrabbit/node-resque/blob/master/lib/plugins/JobLock.js#L4-L5) and modify `beforePerform`. 
  - Explanation: this function is trying to [SETNX](https://redis.io/commands/setnx) the `key`. If set is not successful, that means it's already in the queue, so re-enqueue. In our case, we use `buildId` as args, which is always unique, so it will never re-enqueue. 
  - Beside, if we use the queue to check, that means we will re-enqueue B if A is currently **IN QUEUE**. Please correct me if I'm wrong, I don't think that's what we want. We want to re-enqueue B if A is currently **RUNNING**.  
- So I think we need to keep track of all `jobId` that are running in a table called `runningJobs`. In `beforePerform`, we check if **the blocking job** is running by checking if the `jobId` is in the table. If it is then re-enqueue. If not then we add the **current** `jobId` to table, same for regular case that is not blocked_by. In `afterPerform`, we remove the **current** `jobId`.
- Inside the worker `start`, we will add this new plugin like [this example](https://github.com/taskrabbit/node-resque/blob/1b8509651bf5962b987cce2de819cd1d2c9576b7/examples/example.js#L27-L32)

**TO DO:**
- Data-schema: accept `blocked_by`
- Config-parser: parse `blocked_by`
- Models: pass `blocked_by` to executor `start`
- Executor-queue: make sure the field is passed to worker job `start`
- Worker: implement the new plugin and add it to `start`

**NOTES:** 
This implementation relies on redis and queue, just like `build periodically`.

## Second way (simpler):
- Add a new table called `queuedJobs`. In `start`, add `jobId` to the table. In `stop`, remove from table. Same place where we are adding/removing keys from `buildConfig` table: https://github.com/screwdriver-cd/executor-queue/blob/master/index.js#L279-L281
- Inside `start`, we check if A is in the `queuedJobs` table. If it is, `re-enqueue` B. If not, then call `enqueue` (current flow).

**TO DO:**
- Data-schema: accept `blocked_by`
- Config-parser: parse `blocked_by`
- Models: pass `blocked_by` to executor `start`
- Executor-queue: check against new table and call re-enqueue if necessary

**NOTES:** 
- This implementation relies on redis and queue, just like `build periodically`.
- This implementation is simpler, but it actually checks if A is in the queue, not if A is actually running (since the add/remove from table is being done at executor level). This might be acceptable if the queue is always FIFO and doesn't have prioritization. 
- To solve the above concern, we can also make API call to check the status of latest build of job A instead having an additional table. The code might look messier.

**Comment by d2lam** [comment]
## 5/10/2018
We decided to go with **Implementation 1**  because:
- The 2nd implementation has the limitation mentioned above: job B is blocked if A is **IN QUEUED**, not **RUNNING**
- It's not possible to check if job A is is RUNNING here

**Comment by d2lam** [comment]
## 5/14/2018
PRs ready:
- Models: https://github.com/screwdriver-cd/models/pull/244
- Executor-queue: https://github.com/screwdriver-cd/executor-queue/pull/24

**Comment by d2lam** [comment]
## 5/15/2018
WIP for worker: https://github.com/screwdriver-cd/queue-worker/pull/60

**Comment by d2lam** [comment]
### 5/17/2018
`worker` requires a bigger change because we need to pull in a new **MAJOR** version of `node-resque`:
- We are behind since last year (2017)
- All new docs point to the new version. Old version is deprecated and not supported. We need to write a new plugin, so it's the same amount of effort looking up old code/structure. We might as well pull in the new version.  

The new version changes all the ways things are constructed. So lots of tests & code need to be changed. It is still now _WIP_.

**Comment by d2lam** [comment]
5/21/2018
- Merged all PRs
- Tested on on Beta, seems to work propertly

**Comment by d2lam** [comment]
5/23/2018
Addressed edge case: 
- User starts A1
- User starts A2 - get blocked by itself -> reenqueue in 2 mins
- A1 finishes 
- User starts A3 - not blocked -> start A3
- 2 mins passed. A2 is enqueued -> start A2

Therefore, the events will be out of order: A1->A3->A2. 
This is resolved in https://github.com/screwdriver-cd/queue-worker/pull/63

**Comment by d2lam** [comment]
## 5/25/2018
Fix bug where plugin options not being passed: https://github.com/screwdriver-cd/queue-worker/pull/64

**Comment by tkyi** [comment]
## 5/29/2018
### The Issue
I noticed that it's kind of confusing for users that `blockedBy` has similar syntax to `requires` but we don't enforce use of the tilde(`~`) for internal jobs. Moreover, all `blockedBy` jobs should use OR logic, which we specified as requiring a ~ in our [workflow documentation](https://docs.screwdriver.cd/user-guide/configuration/workflow#advanced-logic).

**Current implementation:**
```
jobs:
      job1:
            requires: [~pr, ~commit]
      job2:
            requires: [~pr, ~commit]
            blockedBy: [job1, ~sd@456:publish]

```
### The Proposal
After discussing with @jithin1987 and @minz1027 , we think it makes sense to require a tilde for all jobs listed for `blockedBy`, like below. This at least makes it clear to users that `blockedBy` is only using OR logic. This also makes it simple to distinguish between OR logic and AND logic if we choose to implement that in the future.
```
blockedBy: [~job1, ~sd@456:publish]
```
We will make a change to the `requires` logic to match this syntax in future work.

**Comment by d2lam** [comment]
## 6/5/2018
Backend is working. 
Need to improve the UI so that user knows the job is blocked. Especially for the case where job is blocked by itself, and user keeps trying to start the same job. It will make our queue really big.
Need to update build status to `BLOCKED` so that UI can display it.

**Comment by tkyi** [comment]
## 6/5/2018
PR opened to display a BLOCKED icon for blocked jobs. Awaiting reviews.

**Comment by d2lam** [comment]
6/19/2018
PR for blocked status:
- https://github.com/screwdriver-cd/queue-worker/pull/80
- https://github.com/screwdriver-cd/screwdriver/pull/1129
- https://github.com/screwdriver-cd/ui/pull/313

**Comment by d2lam** [comment]
This should now be working:
<img width="199" alt="screen shot 2018-06-21 at 2 43 21 pm" src="https://user-images.githubusercontent.com/3401924/41748547-3a241200-7567-11e8-9edd-f6a6d3b821aa.png">

Example pipeline: https://cd.screwdriver.cd/pipelines/1107
Example repo: https://github.com/screwdriver-cd-test/workflow-blockedby-example
